File: day-03-lobby/joltage.php
<?php

echo array_reduce(
    file("input.txt"),
    fn($sum, $line) => $sum += getMax(trim($line), 12)
);

function getMax(string $line, $length = 2): int
{
    list($max, $r) = find($line, $length - 1);

    for($i = $length - 2; $i >= 0; $i--) {
        list($m, $r) = find($r, $i);
        $max .= $m;
    }

    return (int)$max;
}

function find(string $line, $ignore = 0): array
{
    $max = max(str_split($ignore > 0 ? substr($line, 0, -1 * $ignore) : $line));
    $rest = substr($line, strpos($line, $max) + 1);
    return [$max, $rest];
}

File: day-03-lobby/joltage.py
def find(line:str, ignore:int = 0):
    m = max(list(line[:-ignore]) if ignore else line)
    return str(m), line[line.index(m) + 1:]

def power(line:str, length:int = 2):
    power, rest = find(line, length - 1)
    for i in range(length - 2, -1, -1):
        digit, rest = find(rest, i)
        power += digit
    return int(power)

star1, star2 = 0, 0

for battery in open(0).read().splitlines():
    star1 += power(battery, 2)
    star2 += power(battery, 12)

print(f'Star1: {star1}\nStar2: {star2}')